cmd/derper/derper.go

```diff
@@ -200,82 +210,91 @@ func main() {
+	if err != nil {
+		log.Fatalf("derper: can not start cert provider: %v", err)
 	}
 
-	if serveTLS {
-		log.Printf("derper: serving on %s with TLS", *addr)
-		var certManager certProvider
-		certManager, err = certProviderByCertMode(*certMode, *certDir, *hostname)
-		if err != nil {
-			log.Fatalf("derper: can not start cert provider: %v", err)
-		}
-		httpsrv.TLSConfig = certManager.TLSConfig()
-		getCert := httpsrv.TLSConfig.GetCertificate
-		httpsrv.TLSConfig.GetCertificate = func(hi *tls.ClientHelloInfo) (*tls.Certificate, error) {
-			cert, err := getCert(hi)
-			if err != nil {
-				return nil, err
-			}
-			cert.Certificate = append(cert.Certificate, s.MetaCert())
-			return cert, nil
-		}
-		httpsrv.Handler = http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
-			// Set HTTP headers to appease automated security scanners.
-			//
-			// Security automation gets cranky when HTTPS sites don't
-			// set HSTS, and when they don't specify a content
-			// security policy for XSS mitigation.
-			//
-			// DERP's HTTP interface is only ever used for debug
-			// access (for which trivial safe policies work just
-			// fine), and by DERP clients which don't obey any of
-			// these browser-centric headers anyway.
-			w.Header().Set("Strict-Transport-Security", "max-age=63072000; includeSubDomains")
-			w.Header().Set("Content-Security-Policy", "default-src 'none'; frame-ancestors 'none'; form-action 'none'; base-uri 'self'; block-all-mixed-content; plugin-types 'none'")
-			mux.ServeHTTP(w, r)
-		})
+	if *insecurePort > 0 {
 		go func() {
-			port80srv := &http.Server{
-				Addr:        net.JoinHostPort(listenHost, "80"),
-				Handler:     certManager.HTTPHandler(tsweb.Port80Handler{Main: mux}),
+			var handler http.Handler = tsweb.Port80Handler{Main: mux}
+			if *insecurePort == 80 && *certMode == "letsencrypt" {
+				handler = certManager.HTTPHandler(tsweb.Port80Handler{Main: mux})
+			}
+
+			insecureSrv := &http.Server{
+				Addr:        net.JoinHostPort(bindAddress, fmt.Sprintf("%d", *insecurePort)),
+				Handler:     handler,
 				ReadTimeout: 30 * time.Second,
 				// Crank up WriteTimeout a bit more than usually
 				// necessary just so we can do long CPU profiles
 				// and not hit net/http/pprof's "profile
 				// duration exceeds server's WriteTimeout".
 				WriteTimeout: 5 * time.Minute,
 			}
-			err := port80srv.ListenAndServe()
-			if err != nil {
-				if err != http.ErrServerClosed {
-					log.Fatal(err)
-				}
+
+			log.Printf("derper: listening unencrypted on %s", insecureSrv.Addr)
+			err := insecureSrv.ListenAndServe()
+			if err != nil && err != http.ErrServerClosed {
+				log.Fatalf("derper: %v", err)
 			}
 		}()
-		err = httpsrv.ListenAndServeTLS("", "")
-	} else {
-		log.Printf("derper: serving on %s", *addr)
-		err = httpsrv.ListenAndServe()
 	}
-	if err != nil && err != http.ErrServerClosed {
-		log.Fatalf("derper: %v", err)
+
+	if securePort > 0 {
```

Comment:
this whole block gets very indented fast

can this PR do less refactoring? Or can we do refactoring in its own no-op commit that's purely about refactoring?